doc/dvsource-dvgrab.1
.\" dvsource-dvgrab.1 written by Ben Hutchings <ben@decadent.org.uk>
.TH DVSOURCE-DVGRAB 1 "24 May 2009"
.SH NAME
dvsource-dvgrab \- Firewire and Video4Linux2 sources for DVswitch
.SH SYNOPSIS
.HP
.B dvsource-firewire
.RI [ OPTIONS ]
.RB [ \-c
.IR CARD-NUMBER " | " DEVICE ]
.HP
.B dvsource-v4l2-dv
.RI [ OPTIONS "] [" DEVICE ]
.HP
.B dvsource-command
.RI "[ '" FFMPEG_COMMAND "' ]"
.SH DESCRIPTION
.LP
\fBdvsource-firewire\fR uses \fBdvgrab\fR(1) to stream from a DV
device (camera or VCR) connected by Firewire.
.LP
\fBdvsource-v4l2-dv\fR uses \fBdvgrab\fR(1) to stream from a DV device
that has a Video4Linux2 driver.  This includes most USB DV cameras if
you have the uvcvideo driver installed.
.LP
\fBdvsource-command\fR uses a user provided command to create a virtual DV stream
from any device. The command may consist of multiple arguments, if so it must be quoted
as a single shell argument. The default command takes audio from the first
soundcard (builtin microphone) and video from the first video device.
If your webcam is \fB/dev/video0\fR and has its own mono microphone, this
command might work:
.br
\fB'ffmpeg -v quiet -f alsa -ac 1 -i hw:1 -f v4l2 -i /dev/video0 -target pal-dv pipe:1'\fR
.SH OPTIONS
\fB\-t\fR, \fB\-\-tally\fR
.RS
Print notices of activation and deactivation in the mixer, for use
with a tally light.  These will take the form "TALLY: on" or "TALLY:
off".
.RE
.TP
\fB\-v\fR, \fB\-\-verbose\fR
.RS
Print more information at startup.
.RE
.TP
\fB\-h\fR, \fB\-\-host=\fIHOST\fR
.TP
\fB\-p\fR, \fB\-\-port=\fIPORT\fR
.RS
Specify the network address on which DVswitch is listening.  The host
address may be specified by name or as an IPv4 or IPv6 literal.
.RE
.TP
\fB\-c\fR, \fB\-\-card=\fICARD-NUMBER\fR
.RS
Specify the index of the Firewire card the DV device is connected to.
.RE
.TP
\fB-g\fR, \fB\-\-guid=\fIGUID\fR
.RS
Specify the GUID of the Firewire DV device.
.RE
.SH AUTHOR
Ben Hutchings <ben@decadent.org.uk>.
.SH SEE ALSO
dvgrab(1), /usr/share/doc/dvswitch/README


ChangeLog
dvswitch (0.9.1)
  [ Robin Gareus ]
  * OSC remote-control (see doc/README)
  * add contrib examples and tools for OSC
  * menu-entry to toggle title-safe-area-highlight
  * expose status-bar mode (on,off,blink) in Menu.
  * dvsource-file: allow to read from STDIN

  [ James Bromberger ]
  * Add support for printing timing information to screen during
    playout in dvsource-file

dvswitch (0.9) unstable; urgency=low

  [ Wouter Verhelst ]
  * Don't try to build or install dvsource-alsa on non-Linux.
  * Revisit the interface to make it slihgtly more intuitive again.

  [ Robin Gareus ]
  * fast-fades (allow fade effect 40ms..15sec)
  * ffmpeg API update (for ffmpeg-0.10)
  * dynamic A/B overlay
  * confirm application quit dialog
  * blinking status bar
  * updated GUI widget alignment, and labels
  * various updates to Debian packaging
  * various spelling and formating fixes (doc, man-pages)

 -- Wouter Verhelst <wouter@debian.org>  Wed, 28 Mar 2012 00:39:04 +0200

dvswitch (0.9~rc1) experimental; urgency=low

  [ Ben Hutchings ]
  * Added dvsource-alsa as an audio-only source
  * Create one encoding thread per processor, up to a maximum of 8
  * Add status overlay to make it more clear whether recording is enabled
  * Set display to black before a source is connected
  * Add static Makefile at top level
  * Override source frame aspect ratio in case of a mismatch, rather then
    warning about it - Alioth feature #311953
  * Fix APTs and validity flags in header block - Alioth bug #311836
  * List all needed libraries to the linker.  We were accidentally
    depending on the behaviour of GNU ld, which treats 'needed' as
    transitive by default - Debian bug #554310
  * Remove support for 44.1 kHz audio
  * Add update-po target for updating translation files
  * Fix build failure with ffmpeg 0.6
  * Fix crash at startup with ffmpeg 0.6 due to a new parameter check
  * Enable selection of Firewire DV devices by GUID, thanks to
    Manuel Virgilio
  * Implement dvsource-firewire-rtsp, to serve DV over RTSP (experimental)

  [ Carl Karsten ]
  * Add peak level indicators to VU-meter

  [ Wouter Verhelst ]
  * Add support for tally lights in dvsource-firewire/dvsource-v4l2-dv
  * Internationalize and add Dutch translation
  * Disable effects buttons until sufficient sources have connected.
    Pressing them before sources are connected would result in a crash.
  * Implement fading on an 8-bit scale
  * Disable record button if we can't record - Alioth bug #312329
  * Add tally light support to dvsource-firewire-rtsp
  * Install dvsource-firewire-rtsp into .../bin
  * Take over day-to-day maintenance from Ben

  [ Robin Gareus ]
  * Add dvsource-jack, similar to dvsource-alsa but for JACK audio sources

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 01 Mar 2009 03:16:30 +0000

dvswitch (0.8) unstable; urgency=low

  * Added ticks and labels to the VU-meter
  * Bug fixes:
    - Removed use of function not present in Gtkmm 2.8
    - Fixed warnings from CMake 2.6
    - Fixed warnings from GCC
    - Fixed drag-selection when the pointer moves beyond the left or top
      edge of the display
    - Fixed memory corruption/crash in a selection corner-case
    - Fixed synchronisation of display buffers with the X server
    - Fixed repainting of main display before any frames have been received

 -- Ben Hutchings <ben@decadent.org.uk>  Wed, 04 Feb 2009 23:16:13 +0000

dvswitch (0.8~rc1) unstable; urgency=low

  * Disabled recording at startup
  * Added menu, currently with Quit command
  * Added buttons to GUI for effect selection
  * Changed key commands slightly
  * Allowed changing of primary and secondary video sources without
    cancelling the effect
  * Made dvsink-files create directories as necessary
  * Made dvsource-file reject non-DV files
  * Added support for USB DV sources through dvgrab and the uvcvideo driver
  * Moved source selector buttons to keep window within 1024x768 (depending
    on font selection and window decoration)
  * Added VU-meter for basic audio monitoring
  * Added warning indicator for sources which don't match the selected video
    and audio format
  * Documentation changes:
    - Documented all key commands
    - Changed contact address to dvswitch-devel@lists.alioth.debian.org
    - Added notes to (maybe) avoid user mistakes
  * Build system changes:
    - Changed to use cmake
    - Fixed building with current ffmpeg libavcodec library
  * Bug fixes:
    - Changed picture-in-picture selection to maintain frame aspect ratio
    - Restricted picture-in-picture to use the active area of the reduced
      frame
    - Slightly improved colour accuracy of picture-in-picture
    - Fixed cancellation of picture-in-picture selection
    - Fixed crash when user tries to apply picture-in-picture with an empty
      selection
    - Fixed crash when display depth is not 24 or 32
    - Fixed crash when users click in the dvswitch window without any source
      enabled
    - Made dvsource-file work with non-seekable files such as pipes

 -- Ben Hutchings <ben@decadent.org.uk>  Tue, 27 Jan 2009 01:04:43 +0000

dvswitch (0.7) unstable; urgency=low

  * Made keyboard shortcut for cut button work
  * Moved video/DV system parameters into a common structure
  * Replaced libdv codec with DV codec in libavcodec (ffmpeg), which
    is significantly faster
  * Added UI for picture-in-picture mixing

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 20 Apr 2008 22:18:37 +0100

dvswitch (0.6.2) unstable; urgency=low

  * Fixed bug in thumbnail image size initialisation that could result
    in a fatal BadValue error from X
  * Fixed minor bug in thumbnail image cleanup
  * Fixed bug that caused all sinks to be disabled when not recording

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 24 Feb 2008 10:55:41 +0100

dvswitch (0.6.1) unstable; urgency=low

  * Improved X driver compatibility:
    - Fixed calculation of Xv port numbers

src/dvsource-dvgrab.c
/* Copyright 2007-2009 Ben Hutchings.
 * Copyright 2008 Petter Reinholdtsen.
 * Copyright 2009 Wouter Verhelst.
 * See the file "COPYING" for licence details.
 */
/* Source that reads DIF over USB or Firewire using dvgrab */

#include <assert.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <getopt.h>
#include <unistd.h>

#include "config.h"
#include "protocol.h"
#include "socket.h"

static struct option options[] = {
    {"card",     1, NULL, 'c'},
    {"guid",	  1, NULL, 'g'},
    {"firewire", 0, NULL, 'F'},
    {"v4l2",     0, NULL, 'V'},
    {"command",  0, NULL, 'C'},
    {"host",     1, NULL, 'h'},
    {"port",     1, NULL, 'p'},
    {"tally",    0, NULL, 't'},
    {"verbose",  0, NULL, 'v'},
    {"help",     0, NULL, 'H'},
    {NULL,       0, NULL, 0}
};

enum mode {
    mode_unknown,
    mode_firewire,
    mode_v4l2,
    mode_command,
};

static enum mode mode = mode_unknown;
static char * device_name = NULL;
static char * firewire_card = NULL;
static char * firewire_guid = NULL;
static char * mixer_host = NULL;
static char * mixer_port = NULL;
static int do_tally = 0;
static int verbose = 0;

static enum mode program_mode(const char * progname)
{
    const char * slash = strrchr(progname, '/');

    if (slash)
	progname = slash + 1;

    if (strcmp(progname, "dvsource-firewire") == 0)
	return mode_firewire;
    if (strcmp(progname, "dvsource-v4l2-dv") == 0)
	return mode_v4l2;
    if (strcmp(progname, "dvsource-command") == 0)
	return mode_command;
    return mode_unknown;
}

static void handle_config(const char * name, const char * value)
{
    if (strcmp(name, "FIREWIRE_CARD") == 0)
    {
	free(firewire_card);
	firewire_card = strdup(value);
    }
    else if (strcmp(name, "FIREWIRE_DEVICE") == 0 && mode == mode_firewire)
    {
	free(device_name);
	device_name = strdup(value);
    }
    else if (strcmp(name, "V4L2_DV_DEVICE") == 0 && mode == mode_v4l2)
    {
	free(device_name);
	device_name = strdup(value);
    }
    else if (strcmp(name, "MIXER_HOST") == 0)
    {
	free(mixer_host);
	mixer_host = strdup(value);
    }
    else if (strcmp(name, "MIXER_PORT") == 0)
    {
	free(mixer_port);
	mixer_port = strdup(value);
    }
}

static void usage(const char * progname)
{
    static const char firewire_args[] =
	"[-c CARD-NUMBER | DEVICE] [-g GUID]";
    static const char v4l2_args[] = "[DEVICE]";
    static const char cmd_args[] = "['ffmpeg -v quiet -f alsa -ac 1 -i hw:1 -f v4l2 -i /dev/video0 -target pal-dv pipe:1']";
    static const char other_args[] =
	"[-t] [-v] [-h HOST] [-p PORT]";

    switch (program_mode(progname))
    {
    case mode_unknown:
	fprintf(stderr,
		"Usage: %s %s \\\n"
		"           --firewire %s\n"
		"       %s %s \\\n"
		"           --v4l2 %s\n"
		"       %s %s \\\n"
		"           --command %s\n",
		progname, other_args, firewire_args,
		progname, other_args, v4l2_args,
		progname, other_args, cmd_args);
	break;
    case mode_firewire:
	fprintf(stderr,
		"Usage: %s %s \\\n"
		"           %s\n",
		progname, other_args, firewire_args);
	break;
    case mode_v4l2:
	fprintf(stderr,
		"Usage: %s %s \\\n"
		"           %s\n",
		progname, other_args, v4l2_args);
	break;
    case mode_command:
	fprintf(stderr,
		"Usage: %s %s \\\n"
		"           %s\n",
		progname, other_args, cmd_args);
	break;
    }
}

static ssize_t read_retry(int fd, void * buf, size_t count)
{
    ssize_t chunk, total = 0;

    do
    {
        chunk = read(fd, buf, count);
        if (chunk <= 0)
        {
            if (total == 0)
                return chunk;
            break;
        }
        total += chunk;
        buf = (char *)buf + chunk;
        count -= chunk;
    }
    while (count);

    return total;
}

static void tally(int sock)
{
    // Messages should never be buffered
    setbuf(stdout, NULL);

    for (;;)
    {
	char act_msg[ACT_MSG_SIZE];
	ssize_t read_size = read_retry(sock, act_msg, ACT_MSG_SIZE);
	if (read_size < ACT_MSG_SIZE)
	{
	    if (read_size < 0)
		perror("ERROR: read");
	    break;
	}

	if (act_msg[ACT_MSG_VIDEO_POS])
	    printf("TALLY: on\n");
	else
	    printf("TALLY: off\n");
    }

}

int main(int argc, char ** argv)
{
    mode = program_mode(argv[0]);

    /* Initialise settings from configuration files. */
    dvswitch_read_config(handle_config);

    /* Parse arguments. */

    int opt;
    while ((opt = getopt_long(argc, argv, "c:g:h:p:tvFVC", options, NULL)) != -1)
    {
	switch (opt)
	{
	case 'c':
	    free(firewire_card);
	    firewire_card = strdup(optarg);
	    break;
	case 'g':
	    free(firewire_guid);
	    firewire_guid = strdup(optarg);
	    break;
	case 'F':
	    mode = mode_firewire;
	    break;
	case 'V':
	    mode = mode_v4l2;
	    break;
	case 'C':
	    mode = mode_command;
	    break;
	case 'h':
	    free(mixer_host);
	    mixer_host = strdup(optarg);
	    break;
	case 'p':
	    free(mixer_port);
	    mixer_port = strdup(optarg);
	    break;
	case 't':
	    do_tally = 1;
	    break;
	case 'v':
	    verbose = 1;
	    break;
	case 'H': /* --help */
	    usage(argv[0]);
	    return 0;
	default:
	    usage(argv[0]);
	    return 2;
	}
    }

    if (optind != argc)
    {
	free(device_name);
	device_name = strdup(argv[optind++]);
    }

    if (!mixer_host || !mixer_port)
    {
	fprintf(stderr, "%s: mixer hostname and port not defined\n",
		argv[0]);
	return 2;
    }

    if (optind != argc)
    {
	fprintf(stderr, "%s: excess argument \"%s\"\n",
		argv[0], argv[optind]);
	usage(argv[0]);
	return 2;
    }

    if (mode == mode_firewire)
    {
	if (device_name)
	    printf("INFO: Reading from Firewire device %s\n", device_name);
	else if (firewire_card)
	    printf("INFO: Reading from Firewire card %s\n", firewire_card);
	else
	    printf("INFO: Reading from first Firewire card with camera\n");
    }
    else if (mode == mode_v4l2)
    {
	if (!device_name)
	    device_name = "/dev/video";
	printf("INFO: Reading from V4L2 device %s (DV mode)\n", device_name);
    }
    else if (mode == mode_command)
    {
        // FIXME: would it work without audio?
	if (!device_name)
	    device_name = "ffmpeg -v quiet -f alsa -ac 2 -i hw:0 -f v4l2 -i /dev/video0 -target pal-dv pipe:1";
	printf("INFO: Running command: %s\n", device_name);
    }
    else
    {
	fprintf(stderr, "%s: mode not defined (Firewire, V4L2, or Command)\n", argv[0]);
	return 2;
    }

    /* Connect to the mixer. */

    printf("INFO: Connecting to %s:%s\n", mixer_host, mixer_port);
    int sock = create_connected_socket(mixer_host, mixer_port);
    assert(sock >= 0); /* create_connected_socket() should handle errors */
    if (write(sock, do_tally ? GREETING_ACT_SOURCE : GREETING_SOURCE,
              GREETING_SIZE) != GREETING_SIZE)
    {
	perror("ERROR: write");
	exit(1);
    }
    printf("INFO: Connected\n");

    if (do_tally)
    {
	fflush(NULL);
	int child_pid = fork();

	if (child_pid < 0)
	{
	    perror("ERROR: fork");
	    return 1;
	}
	if (child_pid == 0)
	{
	    tally(sock);
	    _exit(0);
	}
    }

    if (mode == mode_command)
      {
        // run a command, e.g. ffmpeg
	if (verbose)
	{
	    printf("INFO: Running %s\n", device_name);
	}

	if (dup2(sock, STDOUT_FILENO) < 0)
	{
	    perror("ERROR: dup2");
	    return 1;
	}
	close(sock);

	execlp("/bin/sh", "/bin/sh", "-c", device_name, NULL);
	perror("ERROR: execlp");
	return 1;
      }

    /* Run dvgrab with the socket as stdout. */

    char * dvgrab_argv[9];
    char ** argp = dvgrab_argv;
    *argp++ = "dvgrab";
    if (mode == mode_v4l2)
	*argp++ = "-v4l2";
    if (device_name)
    {
	*argp++ = "-input";
	*argp++ = device_name;
    }
    else if (firewire_card)
    {
	*argp++ = "-card";
	*argp++ = firewire_card;
    }
    if (firewire_guid)
    {
	*argp++ = "-guid";
	*argp++ = firewire_guid;
    }
    *argp++ = "-noavc";
    *argp++ = "-";
    *argp = NULL;
    assert(argp < dvgrab_argv + sizeof(dvgrab_argv) / sizeof(dvgrab_argv[0]));

    if (verbose)
    {
	printf("INFO: Running");
	for (argp = dvgrab_argv; *argp != NULL; argp++)
	    printf(" %s", *argp);
	printf("\n");
    }

    if (dup2(sock, STDOUT_FILENO) < 0)
    {
	perror("ERROR: dup2");
	return 1;
    }
    close(sock);

    execvp("dvgrab", dvgrab_argv);
    perror("ERROR: execvp");
    return 1;
}
